File: app/src/main/java/com/kimjisub/launchpad/midi/driver/LaunchpadPRO.java
package com.kimjisub.launchpad.midi.driver;

import com.kimjisub.manager.Log;

public class LaunchpadPRO extends DriverRef {
	static final int[][] circleCode = {
			{11, -80, 91},
			{11, -80, 92},
			{11, -80, 93},
			{11, -80, 94},
			{11, -80, 95},
			{11, -80, 96},
			{11, -80, 97},
			{11, -80, 98},
			{11, -80, 89},
			{11, -80, 79},
			{11, -80, 69},
			{11, -80, 59},
			{11, -80, 49},
			{11, -80, 39},
			{11, -80, 29},
			{11, -80, 19},
			{11, -80, 8},
			{11, -80, 7},
			{11, -80, 6},
			{11, -80, 5},
			{11, -80, 4},
			{11, -80, 3},
			{11, -80, 2},
			{11, -80, 1},
			{11, -80, 10},
			{11, -80, 20},
			{11, -80, 30},
			{11, -80, 40},
			{11, -80, 50},
			{11, -80, 60},
			{11, -80, 70},
			{11, -80, 80}
	};

	@Override
	public void getSignal(int cmd, int sig, int note, int velo) {
		if (cmd == 9) {
			int x = 9 - (note / 10);
			int y = note % 10;

			if (y >= 1 && y <= 8)
				onPadTouch(x - 1, y - 1, velo != 0, velo);
		}
		if (cmd == 11 && sig == -80) {
			if (91 <= note && note <= 98) {
				onFunctionkeyTouch(note - 91, velo != 0);
			}
			if (19 <= note && note <= 89 && note % 10 == 9) {
				int c = 9 - (note / 10) - 1;
				onChainTouch(c, velo != 0);
				onFunctionkeyTouch(c + 8, velo != 0);
			}
			if (1 <= note && note <= 8) {
				onChainTouch(8 - note + 16 - 8, velo != 0);
				onFunctionkeyTouch(8 - note + 16, velo != 0);
			}
			if (10 <= note && note <= 80 && note % 10 == 0) {
				onChainTouch((note / 10) - 1 + 24 - 8, velo != 0);
				onFunctionkeyTouch((note / 10) - 1 + 24, velo != 0);
			}

		} else {
			onUnknownEvent(cmd, sig, note, velo);

			if (cmd == 7 && sig == 46 && note == 0 && velo == -9)
				Log.midiDetail("PRO > Live Mode");
			else if (cmd == 23 && sig == 47 && note == 0 && velo == -9)
				Log.midiDetail("PRO > Note Mode");
			else if (cmd == 23 && sig == 47 && note == 1 && velo == -9)
				Log.midiDetail("PRO > Drum Mode");
			else if (cmd == 23 && sig == 47 && note == 2 && velo == -9)
				Log.midiDetail("PRO > Fade Mode");
			else if (cmd == 23 && sig == 47 && note == 3 && velo == -9)
				Log.midiDetail("PRO > Programmer Mode");
		}
	}

	@Override
	public void sendPadLED(int x, int y, int velo) {
		send09Signal(10 * (8 - x) + y + 1, velo);
	}

	@Override
	public void sendChainLED(int c, int velo) {
		if (0 <= c && c <= 7)
			sendFunctionkeyLED(c + 8, velo);
	}

	@Override
	public void sendFunctionkeyLED(int f, int velo) {
		if (0 <= f && f <= 31)
			sendSignal((byte) circleCode[f][0], (byte) circleCode[f][1], (byte) circleCode[f][2], (byte) velo);
	}


	@Override
	public void sendClearLED() {
		for (int i = 0; i < 8; i++)
			for (int j = 0; j < 8; j++)
				sendPadLED(i, j, 0);
		for (int i = 0; i < 32; i++)
			sendFunctionkeyLED(i, 0);
	}
}

File: app/src/main/java/com/kimjisub/launchpad/midi/driver/LaunchpadPRO.kt
package com.kimjisub.launchpad.midi.driver

import com.kimjisub.launchpad.tool.Log

class LaunchpadPRO : DriverRef() {
	companion object {
		internal val circleCode = arrayOf(
			intArrayOf(11, -80, 91),
			intArrayOf(11, -80, 92),
			intArrayOf(11, -80, 93),
			intArrayOf(11, -80, 94),
			intArrayOf(11, -80, 95),
			intArrayOf(11, -80, 96),
			intArrayOf(11, -80, 97),
			intArrayOf(11, -80, 98),
			intArrayOf(11, -80, 89),
			intArrayOf(11, -80, 79),
			intArrayOf(11, -80, 69),
			intArrayOf(11, -80, 59),
			intArrayOf(11, -80, 49),
			intArrayOf(11, -80, 39),
			intArrayOf(11, -80, 29),
			intArrayOf(11, -80, 19),
			intArrayOf(11, -80, 8),
			intArrayOf(11, -80, 7),
			intArrayOf(11, -80, 6),
			intArrayOf(11, -80, 5),
			intArrayOf(11, -80, 4),
			intArrayOf(11, -80, 3),
			intArrayOf(11, -80, 2),
			intArrayOf(11, -80, 1),
			intArrayOf(11, -80, 10),
			intArrayOf(11, -80, 20),
			intArrayOf(11, -80, 30),
			intArrayOf(11, -80, 40),
			intArrayOf(11, -80, 50),
			intArrayOf(11, -80, 60),
			intArrayOf(11, -80, 70),
			intArrayOf(11, -80, 80)
		)
	}

	override fun getSignal(cmd: Int, sig: Int, note: Int, velocity: Int) {
		if (cmd == 9) {
			val x = 9 - note / 10
			val y = note % 10
			if (y in 1..8)
				onPadTouch(x - 1, y - 1, velocity != 0, velocity)
		}
		if (cmd == 11 && sig == -80) {
			if (note in 91..98) {
				onFunctionKeyTouch(note - 91, velocity != 0)
			}
			if (note in 19..89 && note % 10 == 9) {
				val c = 9 - note / 10 - 1
				onChainTouch(c, velocity != 0)
				onFunctionKeyTouch(c + 8, velocity != 0)
			}
			if (note in 1..8) {
				onChainTouch(8 - note + 16 - 8, velocity != 0)
				onFunctionKeyTouch(8 - note + 16, velocity != 0)
			}
			if (note in 10..80 && note % 10 == 0) {
				onChainTouch(note / 10 - 1 + 24 - 8, velocity != 0)
				onFunctionKeyTouch(note / 10 - 1 + 24, velocity != 0)
			}
		} else {
			onUnknownReceived(cmd, sig, note, velocity)
			if (cmd == 7 && sig == 46 && note == 0 && velocity == -9)
				Log.midiDetail("PRO > Live Mode")
			else if (cmd == 23 && sig == 47 && note == 0 && velocity == -9)
				Log.midiDetail("PRO > Note Mode")
			else if (cmd == 23 && sig == 47 && note == 1 && velocity == -9)
				Log.midiDetail("PRO > Drum Mode")
			else if (cmd == 23 && sig == 47 && note == 2 && velocity == -9)
				Log.midiDetail("PRO > Fade Mode")
			else if (cmd == 23 && sig == 47 && note == 3 && velocity == -9)
				Log.midiDetail("PRO > Programmer Mode")
		}
	}

	override fun sendPadLed(x: Int, y: Int, velocity: Int) {
		sendSignal(9, -112, 10 * (8 - x) + y + 1, velocity)
	}

	override fun sendChainLed(c: Int, velocity: Int) {
		if (c in 0..7)
			sendFunctionkeyLed(c + 8, velocity)
	}

	override fun sendFunctionkeyLed(f: Int, velocity: Int) {
		if (f in 0..31)
			sendSignal(
				circleCode[f][0].toByte(),
				circleCode[f][1].toByte(),
				circleCode[f][2].toByte(),
				velocity.toByte()
			)
	}


	override fun sendClearLed() {
		for (i in 0..7)
			for (j in 0..7)
				sendPadLed(i, j, 0)
		for (i in 0..31)
			sendFunctionkeyLed(i, 0)
	}
}